File: src/mvmctl/api/inputs/_key_input.py
"""SSH key input models for API boundary — existing resource actions."""

from __future__ import annotations

from dataclasses import dataclass, field

from mvmctl.core._shared import Database
from mvmctl.core.key._resolver import KeyResolver
from mvmctl.exceptions import KeyNotFoundError
from mvmctl.models import SSHKeyItem

__all__ = [
    "KeyInput",
    "KeyRequest",
    "ResolvedKeyInput",
]


@dataclass
class KeyInput:
    """Input model for identifying existing SSH keys."""

    name: list[str] = field(default_factory=list)
    id: list[str] = field(default_factory=list)


@dataclass(frozen=True)
class ResolvedKeyInput:
    """Immutable resolved key request."""

    keys: list[SSHKeyItem]


class KeyRequest:
    """Resolve key identifiers to DB records."""

    _result: ResolvedKeyInput | None = None

    def __init__(self, *, inputs: KeyInput, db: Database | None = None) -> None:
        self._inputs = inputs
        from mvmctl.core.key._repository import KeyRepository

        self._resolver = KeyResolver(KeyRepository(db))

    @property
    def result(self) -> ResolvedKeyInput | None:
        return self._result

    def resolve(self) -> ResolvedKeyInput:
        identifiers = self._inputs.name + self._inputs.id
        if not identifiers:
            raise KeyNotFoundError("No key identifiers provided")

        result = self._resolver.resolve_many(identifiers)
        if result.errors and not result.items:
            raise KeyNotFoundError(
                f"Could not resolve any keys: {', '.join(result.errors)}"
            )

        self._result = ResolvedKeyInput(keys=result.items)
        return self._result

File: pkg/api/inputs/key_input.go
package inputs

import (
	"context"
	"fmt"
	"strings"

	"mvmctl/internal/core/key"
	"mvmctl/internal/lib/model"
	"mvmctl/pkg/errs"
)

// KeyInput is the raw input for identifying existing SSH keys.
// struct behavior — identifiers are resolved
// by name or ID in a single pass (lumping both).
type KeyInput struct {
	Identifiers []string `json:"identifiers,omitempty"`
}

// Validate checks that the key input has valid identifiers.
func (i *KeyInput) Validate() error {
	if len(i.Identifiers) == 0 {
		return fmt.Errorf("at least one key identifier is required")
	}
	return nil
}

// Resolve resolves all identifiers in the input to SSHKeyItem objects.
// Delegates to key.Resolver.ResolveMany for batch resolution with
// deduplication and error collection.
func (i *KeyInput) Resolve(ctx context.Context, repo key.Repository) ([]*model.SSHKeyItem, error) {
	if err := i.Validate(); err != nil {
		return nil, err
	}
	resolver := key.NewResolver(repo)
	result, err := resolver.ResolveMany(ctx, i.Identifiers)
	if err != nil {
		return nil, err
	}
	if len(result.Errors) > 0 && len(result.Items) == 0 {
		return nil, errs.NotFound(errs.CodeKeyNotFound,
			fmt.Sprintf("failed to resolve keys: %s", strings.Join(result.Errors, "; ")))
	}
	if len(result.Errors) > 0 {
		return result.Items, fmt.Errorf("partial resolve failures: %s", strings.Join(result.Errors, "; "))
	}
	return result.Items, nil
}
